add a cookbook that describes tracking finite strain

jdannberg opened the pull request:
This section describes how to track finite strain using compositional fields. @gassmoeller already agreed to extend it to also include how to model the same process using particles.

bangerth commented on cookbooks/finite_strain/finite_strain.cc:
```diff
@@ -0,0 +1,114 @@
+    evaluate(const MaterialModel::MaterialModelInputs<dim> &in,
+             MaterialModel::MaterialModelOutputs<dim> &out) const
+    {
+      Simple<dim>::evaluate(in, out);
```
Add a comment here that this pre-populates all fields from the Simple model, and that we will only explicitly overwrite a selected set of fields below.

bangerth commented on cookbooks/finite_strain/finite_strain.cc:
```diff
@@ -0,0 +1,114 @@
+    class FiniteStrain : public MaterialModel::Simple<dim>
+    {
+      public:
+
```
not really all that important, but since I'm reviewing -- drop this empty line.

bangerth commented on cookbooks/finite_strain/finite_strain.cc:
```diff
@@ -0,0 +1,114 @@
+    {
+      Simple<dim>::evaluate(in, out);
+
+      // we need the velocity gradiens for the finite strain (they are not included in material model inputs)
```
gradiens -> gradients

bangerth commented on cookbooks/finite_strain/finite_strain.cc:
```diff
@@ -0,0 +1,114 @@
+      Simple<dim>::evaluate(in, out);
+
+      // we need the velocity gradiens for the finite strain (they are not included in material model inputs)
+      // so we get them from the finite element
```
Ah, what a bummer. Creating an `FEValues` object is expensive, but I don't really see a good alternative.

bangerth commented on cookbooks/finite_strain/finite_strain.cc:
```diff
@@ -0,0 +1,114 @@
+                                 update_values   |
+                                 update_gradients |
+                                 update_quadrature_points |
+                                 update_JxW_values);
```
Don't you really only need `update_gradients`?

bangerth commented on cookbooks/finite_strain/finite_strain.cc:
```diff
@@ -0,0 +1,114 @@
+        // existing strain rate tensor components.
+        for (unsigned int q=0; q < in.position.size(); ++q)
+        {
+          // rotation tensor = asymmetric part of the velocity gradients tensor + unit tensor
```
maybe say "asymmetric part of the displacement in this time step (= velocity * time step) + unit tensor"

bangerth commented on cookbooks/finite_strain/finite_strain.cc:
```diff
@@ -0,0 +1,114 @@
+        for (unsigned int q=0; q < in.position.size(); ++q)
+        {
+          // rotation tensor = asymmetric part of the velocity gradients tensor + unit tensor
+          const Tensor<2,dim> rotation = (velocity_gradients[q] - in.strain_rate[q]) * this->get_timestep()
```
It's not wrong to use `in.strain_rate[q]`, but I think it would be easier to read (because the provenance of the variable is clear) to write
```
  velocity_gradients[q] - symmetrize(velocity_gradients[q])
```

bangerth commented on cookbooks/finite_strain/finite_strain.cc:
```diff
@@ -0,0 +1,114 @@
+          const Tensor<2,dim> rotation = (velocity_gradients[q] - in.strain_rate[q]) * this->get_timestep()
+                                         + unit_symmetric_tensor<dim>();
+
+          Tensor<2,dim> strain;
```
maybe call it `accumulated_strain` or `total_strain`?

naliboff replied:
One vote for  `total_strain`.

jdannberg replied:
Oh, I already updated this to `accumulated_strain` (it only shows the old version because this is what Wolfgang's comment was referring to). But if you click below on the commit  `add a cookbook that describes tracking finite strain` or the `Files changed` tab you will see the most recent version (just without the comments).

naliboff replied:
I see it set to `accumulated_strain` that is perfectly clear as well.  Ignore my vote above.

bangerth commented on cookbooks/finite_strain/finite_strain.cc:
```diff
@@ -0,0 +1,114 @@
+    ASPECT_REGISTER_MATERIAL_MODEL(FiniteStrain,
+                                   "finite strain",
+                                   "A simple material model that is like the "
+                                   "'Simple' model, but has a non-zero compressibility "
```
Where is the compressibility actually set?

jdannberg replied:
It is not compressible. I updated the documentation.

bangerth commented on doc/manual/cookbooks/finite_strain/finite_strain.cc:
```diff
@@ -0,0 +1,20 @@
+for (unsigned int q=0; q < in.position.size(); ++q)
+{
```
update this file as necessary once its origin is updated

bangerth commented on doc/manual/manual.tex:
```diff
@@ -5057,6 +5057,52 @@ \subsubsection{Artificial viscosity smoothing}
+In many geophysical settings, material properties, and in particular the rheology, do not only depend
+on the current temperature, pressure and strain rate, but also on the history of the system. 
+This can be incorporated in a geodynamic model by tracking the strain that accumulates over time
+at every point in the model. Tracking this finite strain can either be done by using compositional
```
There are other factors that describe history, so maybe the following would be a better sentence:

"This can be incorporated in \aspect{} models by tracking history variables through compositional fields. In this cookbook, we will show how to do this by tracking the strain that accumulates over time at every (Lagrangian) point in the model."

bangerth replied:
Or something similar that makes it clear that this is only one of the history variables one may want to track.

bangerth commented on doc/manual/manual.tex:
```diff
@@ -5057,6 +5057,52 @@ \subsubsection{Artificial viscosity smoothing}
+The right-hand side term consists of two parts:
+The first one, $R \varepsilon R^T$, rotates the accumulated strain from all the previous time steps with
+the rotation part of the flow field of the current time step, so that strain the material undergoes in the
+current time step can be applied to the correct components. The second part $\dot\varepsilon \Delta t$
```
We've usually written the strain rate as `$\varepsilon(\mathbf u)$`.

jdannberg replied:
Yes, this would be much more consistent with the rest of the manual, but what variable would I use for the strain in this case? Now I have strain = \varepsilon, and strain rate = \dot\varepsilon, which is also what is commonly used in geodynamics, but maybe you have a better idea?

bangerth replied:
That's when one needs to make up new notation :-) Say, 
```
Let us denote the accumulated strain at time step $n$ as $E^n$. We can express it as the sum of the stress $E^{n-1}$ at the previous time step, rotated by the rotational component of the velocity field, plus the strain increment $\varepsilon(\mathbf u^n) \Delta t^n$ accumulated over the current time step.
```

jdannberg replied:
I think it would be nice to also be consistent with the notation that is typically used in geodynamics. If we introduce our own notation, it might be a bit confusing for readers who are used to the variable names that are commonly used. For example, I would associate $E$ with an energy or elastic modulus. 
So I think I would prefer it to change \varepsilon to \dot\varepsilon everywhere in the manual where we refer to the strain rate, and then use \varepsilon for the strain.

bangerth replied:
In our papers, and in the manual, we don't actually use `\varepsilon` as a variable of its own, but always as an operator that simply takes the symmetric gradient. So maybe doing something like `\varepsilon(\mathbf U)` would work, where we say that `\mathbf U` is the displacement vector so that `\partial\mathbf U/\partial t=\mathbf u`? Or maybe use `\mathbf d` for the displacement? Then you could use `\varepsilon(\mathbf d)`.

In principle, I may agree that using `\varepsilon` and `\dot\varepsilon` would be desirable. But that ship has sailed several years ago.

jdannberg replied:
Okay, then I think I would just use $e$ for the strain (which is also used sometimes), and write down how we use these variable names as the very first sentence of the cookbook. Then I can use $\varepsilon(\mathbf u^n)$ for the strain rate, and I hope it is also clear enough that readers do not find it confusing.

naliboff replied:
I found the current definition for strain rate in the cookbook perfectly clear.   I agree, though, that confusion could arise if the the strain rate is defined slightly different in the documentation for a material model, for example.  I have used `\dot{\varepsilon}` to describe strain rate in comments/documentation and this is indeed the common convention in geodynamics.  Should these conventions be synchronized throughout the entire code?

jdannberg replied:
As Wolfgang mentioned, we use `\varepsilon(\mathbf u)` for the strain rate everywhere else in the code, the manual and the first Aspect paper, so to be consistent with all of the other documentation, I changed the variable names: https://github.com/geodynamics/aspect/pull/860/files

naliboff replied:
Yes, keeping things consistent throughout (first paper, manual, code, etc) seems like the best practice.

bangerth commented on doc/manual/manual.tex:
```diff
@@ -5057,6 +5057,52 @@ \subsubsection{Artificial viscosity smoothing}
+The first one, $R \varepsilon R^T$, rotates the accumulated strain from all the previous time steps with
+the rotation part of the flow field of the current time step, so that strain the material undergoes in the
+current time step can be applied to the correct components. The second part $\dot\varepsilon \Delta t$
+adds the strain of the current time step. The new integrated strain is the the sum of the rotated old 
```
the the -> the

bangerth commented on doc/manual/manual.tex:
```diff
@@ -5057,6 +5057,52 @@ \subsubsection{Artificial viscosity smoothing}
+Heating from the bottom leads to the the ascent of plumes from the bottom boundary layer, and the
+associated deformation is visible in the components of the accumulated finite strain. 
+Material moves to the sides at the top of the plume head, so that it is shortened in vertical 
+direction (blue areas in the yy component) and stretched in horizontal direction (red areas in 
```
$yy$ (and $xx$ in the next line, then $xy$ in the one after).

bangerth commented:
This is awesome! Thanks for writing this, it is so short and elegant!

In the section for the manual, can you add a sentence that the example used here only shows how one can integrate up these history variables, but that we do not use them as part of the material model -- but that it would of course be completely trivial to use them in the material model because the values of the compositional fields are part of what the material model gets as inputs.

jdannberg commented:
Okay, I think I have addressed all of your comments now. Thanks for reviewing!

@naliboff, maybe you could also have a look at this? Just to see if all of the definitions for the strain and strain rate make sense (I am never really sure if what we commonly call the 'strain tensor' is only the symmetric part of the gradient of displacements, or the full one).

naliboff commented:
Juliane, this is fantastic and thanks again for putting it together!  The cookbook is quite clear as is the code, although I am still working through what some of the called functions explicitly do.  

I think it would be helpful to have an example where the accumulated strain feeds back into the material model, but perhaps this is something I can help contribute?

Strain and strain-rate definitions: what you have looks correct to me, but admittedly I have not written this out in a good while.  The instantaneous (non-accumulated strain) strain can be separated from the antisymmetric part of the tensor, but are accumulated strain magnitudes correct if the rotation is not taken into account?  I don't think they would be unless one was calculating and adding principal strains.  Maybe I'm over thinking this.   I will go back through the code I used previously to see how it compares.  From memory, though, I believe it used the full gradient of displacement tensor.

A few questions below, some of them naive ;)

1. Is the routine to calculate the strain is being called only after the non-linear iterations are finished?  For example, if the 'iterated Stokes' solver is used, would routine only use the velocity from the final non-linear iteration?  This wasn't clear to me from the source code, but I may have just missed where this is specified.
2. Accessing the accumulated strain in a material model: this would be done through the `reaction_term` variable?
3. Material averaging ... broadly, any big issues that come to mind if one has to average compositions over an element?  Can the reaction terms be averaged in the same manner as all the other values?

naliboff commented on cookbooks/finite_strain/finite_strain.cc:
```diff
@@ -0,0 +1,114 @@
+          for (unsigned int c=0; c<Tensor<2,dim>::n_independent_components; ++c)
+            {
+              out.reaction_terms[q][c] = - in.composition[q][c]
+                                         + rotated_strain[Tensor<2,dim>::unrolled_to_component_indices(c)];
```
Actually, a question about lines 79-80 `out.reaction_terms[q][c]`.  Here you subtract `in.composition[q][c]` so as not to double account the prior accumulated strain.  Should `in.composition[q][c]` (old strain) be the rotated old strain instead?

bangerth replied:
This is correct, though maybe for quirky reasons. In essence, the way ASPECT describes these reaction terms is as an increment from the previous to the new value in Lagrangian coordinates (i.e., ignoring the advection for a moment). What you want is really the difference between the old and the new value, which is what is computed here.

naliboff replied:
Thanks, this makes sense now.

bangerth commented:
@naliboff: Regarding your questions:
1/ Yes, I think it is called only after the iterated Stokes solver completes (or in fact, at the beginning of the next time step, before we get to the Stokes solver). It is a good question what happens if we do the iterated IMPES solver, because then one calls this function multiple times -- I'm not sure anyone has tried this.
2/ No, in the material model you get as inputs the pressure, temperature, strain rate, and compositional values at each quadrature point. The latter contain as a subset the accumulated strains computed here.
3/ See #698.

bangerth commented on doc/manual/manual.tex:
```diff
@@ -5057,6 +5057,67 @@ \subsubsection{Artificial viscosity smoothing}
+\textit{This section was contributed by Juliane Dannberg}
+
+\textit{Notation: Here, we denote the strain rate tensor as $\varepsilon (\mathbf u)$, where 
+$\varepsilon_{ij} = \frac{1}{2}(\frac{\partial u_i}{\partial x_j} + \frac{\partial u_j}{\partial x_i})$ 
```
Can you write this as `$varepsilon(\mathbf u)_{ij}` at the top? You probably want `\left(...\right)` around the two fractions to make it look good.

bangerth commented on doc/manual/manual.tex:
```diff
@@ -5057,6 +5057,67 @@ \subsubsection{Artificial viscosity smoothing}
+\textit{Notation: Here, we denote the strain rate tensor as $\varepsilon (\mathbf u)$, where 
+$\varepsilon_{ij} = \frac{1}{2}(\frac{\partial u_i}{\partial x_j} + \frac{\partial u_j}{\partial x_i})$ 
+with the velocity $\mathbf u$, and the strain tensor as $e (\mathbf d)$, where 
+$e_{ij} = \frac{1}{2}(\frac{\partial d_i}{\partial x_j} + \frac{\partial d_j}{\partial x_i})$ 
```
If you define it this way, you should either write just `e`, or `\varepsilon(\mathbf d)`. Writing `e(d)` unnecessarily introduces `e` as an operator (namely, the operator that had previously already been called `\varepsilon(.)`), where you really just want it to be a tensor.

bangerth commented:
@jdannberg -- this looks good to me with the exception of the two small issues in the manual. Please address as appropriate and merge yourself! Thanks a lot for writing this!

naliboff commented:
@bangerth, @jdannberg -  if the function is called multiple times in a given time step (iterated Stokes, iterated IMPES), that would likely lead to 'artificial' strain accumulation.  Probably could be resolved via a simple if statement?

bangerth commented:
@naliboff -- it would probably be useful to check the following, but I believe that in each iteration of iterated Stokes, we always start from the same "previous solution" but provide iteratively better linearization points for the nonlinear terms. In other words, I don't think we update the solution vector multiple times. We just use the updated vector from the previous iteration to evaluate nonlinear coefficients, but add the updates on top of the previous time step's solution.

jdannberg commented:
@naliboff -- I believe there won't be any problems with using the iterated IMPES scheme. I also use the reaction terms for models with melt migration (where I also have to use the iterated IMPES) and what happens is exactly what Wolfgang described: The input for the material model is updated with every iteration (so that the reaction terms can be computed with the iteratively better solutions), but in the end the reaction term is always added to the solution from the previous time step.

jdannberg commented:
Okay, I fixed the two issues. @bangerth, @naliboff -- thanks a lot for your feedback!
@naliboff -- I could also extend the cookbook with a second part where the accumulated strain feeds back into the material model, but I have never used a material model like this, so it would be great if you could provide a parametrization for how this feedback should look like. 
Alternatively, we would of course be happy if you want to contribute a cookbook where the accumulated strain is used in the material model!

bangerth commented:
For the purpose of a cookbook, the rheology you describe doesn't actually have to be realistic. It would be entirely sufficient to just multiply the viscosity with something like `std::exp(-accumulated_strain/reference_strain)`. The point simply being to show how one can extract the accumulated strain from the material model inputs and use it in the definition of the viscosity.

naliboff commented:
@jdannberg, @bangerth - The strain is typically used to modify brittle parameters and in some cases for viscosity.  For example, as the accumulated strain goes from a value of 0 to 1, the friction angle might decrease linearly from an initial value to a defined 'weakened value'. Similarly, you got reduce the viscosity in a similar manner or through the equation Wolfgang outlined above.  This is next on my to-do list and will definitely go in a cookbook, hopefully ready to review by the end of May. Thanks again Juliane!
